src/lightning/data/streaming/functions.py

```diff
@@ -0,0 +1,188 @@
+
+    if indexed_paths[0] != absolute_path:
+        raise ValueError("The provided path should be absolute.")
+
+    return "/" + os.path.join(*str(absolute_path).split("/")[:4])
+
+
+class LambdaDataTransformRecipe(DataTransformRecipe):
+    def __init__(self, fn: Callable[[str, Any], None], inputs: Sequence[Any]):
+        super().__init__()
+        self._fn = fn
+        self._inputs = inputs
+
+    def prepare_structure(self, input_dir: Optional[str]) -> Any:
+        return self._inputs
+
+    def prepare_item(self, output_dir: str, item_metadata: Any) -> None:  # type: ignore
+        self._fn(output_dir, item_metadata)
+
+
+class LambdaDataChunkRecipe(DataChunkRecipe):
+    def __init__(
+        self,
+        fn: Callable[[Any], None],
+        inputs: Sequence[Any],
+        chunk_size: Optional[int],
+        chunk_bytes: Optional[int],
+        compression: Optional[str],
+    ):
+        super().__init__(chunk_size=chunk_size, chunk_bytes=chunk_bytes, compression=compression)
+        self._fn = fn
+        self._inputs = inputs
+
+    def prepare_structure(self, input_dir: Optional[str]) -> Any:
+        return self._inputs
+
+    def prepare_item(self, item_metadata: Any) -> Any:  # type: ignore
+        if isinstance(self._fn, GeneratorType):
+            yield from self._fn(item_metadata)
+        else:
+            yield self._fn(item_metadata)
+
+
+def map(
+    fn: Callable[[str, Any], None],
+    inputs: Sequence[Any],
+    output_dir: str,
+    num_workers: Optional[int] = None,
+    fast_dev_run: Union[bool, int] = False,
+    num_nodes: Optional[int] = None,
+    machine: Optional[str] = None,
+) -> None:
+    """This function map a callbable over a collection of files possibly in a distributed way.
+
+    Arguments:
+        fn: A function to be executed over each input element
+        inputs: A sequence of input to be processed by the `fn` function.
+            Each input should contain at least a valid filepath.
+        output_dir: The folder where the processed data should be stored.
+        num_workers: The number of workers to use during processing
+        fast_dev_run: Whether to use process only a sub part of the inputs
+        num_nodes: When doing remote execution, the number of nodes to use.
+        machine: When doing remote execution, the machine to use.
+
+    """
+
+    if len(inputs) == 0:
+        raise ValueError(f"The provided inputs should be non empty. Found {inputs}.")
+
+    if num_nodes is None or int(os.getenv("DATA_OPTIMIZER_NUM_NODES", 0)) > 0:
+        remote_output_dir = _LightningSrcResolver()(output_dir)
+
+        if remote_output_dir is None or "cloudspaces" in remote_output_dir:
+            raise ValueError(
+                f"The provided `output_dir` isn't valid. Found {output_dir}."
+                " HINT: You can either use `/teamspace/s3_connections/...` or `/teamspace/datasets/...`."
+            )
+
+        inputs = inputs() if callable(inputs) else inputs
```

Comment:
`inputs` cannot be a callable with the current signature (also this is at odds with the check on `len` above)

Reply:
removed.